server/asset/dcr/dcr.go

```diff
@@ -384,66 +380,128 @@ func unconnectedDCR(logger dex.Logger) *Backend {
+			prevHash, err := chainhash.NewHashFromStr(block.PreviousHash)
+			if err != nil {
+				sendErrFmt("error parsing previous hash %s: %v", block.PreviousHash, err)
+				continue
+			}
+			// If it builds on the best block or the cache is empty, it's good to add.
+			if *prevHash == tip.hash || tip.height == 0 {
+				dcr.log.Debugf("Run: Processing new block %s", bestHash)
+				addBlock(block, false)
+				continue
+			}
+			// It is either a reorg, or the previous block is not the cached
+			// best block. Crawl blocks backwards until finding a mainchain
+			// block, flagging blocks from the cache as orphans along the way.
+			iHash := &tip.hash
+			reorgHeight := int64(0)
+			for {
+				if *iHash == zeroHash {
+					break
+				}
+				iBlock, err := dcr.node.GetBlockVerbose(iHash, false)
 				if err != nil {
-					dcr.log.Errorf("onBlockConnected error retrieving block %s: %v", blockHash, err)
-					return
+					sendErrFmt("error retrieving block %s: %v", iHash, err)
+					break
 				}
-				// Check if this forces a reorg.
-				currentTip := int64(dcr.blockCache.tipHeight())
-				if blockVerbose.Height <= currentTip {
-					dcr.blockCache.reorg(blockVerbose)
+				if iBlock.Confirmations > -1 {
+					// This is a mainchain block, nothing to do.
+					break
 				}
-				block, err := dcr.blockCache.add(blockVerbose)
-				if err != nil {
-					dcr.log.Errorf("error adding block to cache")
+				if iBlock.Height == 0 {
+					break
 				}
-				dcr.signalMtx.RLock()
-				for _, c := range dcr.blockChans {
-					select {
-					case c <- block.height:
-					default:
-						dcr.log.Errorf("tried sending block update on blocking channel")
-					}
+				reorgHeight = iBlock.Height
+				iHash, err = chainhash.NewHashFromStr(iBlock.PreviousHash)
+				if err != nil {
+					sendErrFmt("error decoding previous hash %s for block %s: %v",
+						iBlock.PreviousHash, iHash.String(), err)
+					// Some blocks on the side chain may not be flagged as
+					// orphaned, but still proceed, flagging the ones we have
+					// identified and adding the new best block to the cache and
+					// setting it to the best block in the cache.
+					break
 				}
-				dcr.signalMtx.RUnlock()
-			default:
-				dcr.log.Warn("unknown message type in Run: %T", rawMsg)
 			}
+			var reorg bool
+			if reorgHeight > 0 {
+				reorg = true
+				dcr.log.Infof("Reorg from %s (%d) to %s (%d) detected.",
+					tip.hash, tip.height, bestHash, block.Height)
+				dcr.blockCache.reorg(reorgHeight)
+			}
+			// Now add the new block.
+			addBlock(block, reorg)
+
 		case <-ctx.Done():
 			break out
 		}
 	}
 }
 
-// A callback to be registered with dcrd. It is critical that no RPC calls are
-// made from this method. Doing so will likely result in a deadlock, as per
-// https://github.com/decred/dcrd/blob/952bd7bba34c8aeab86f63f9c9f69fc74ff1a7e1/rpcclient/notify.go#L78
-func (dcr *Backend) onBlockConnected(serializedHeader []byte, _ [][]byte) {
```

Comment:
This comment and other considerations like the buffering being a poor substitute for a real queue are probably good reason to just do polling.  Is that part of the motivation?  I'd like to document the reasoning pretty clearly in the commit message when this is merged.

Reply:
I've added some explanation to the commit message, but I'll summarize here. The major factor for me was simplicity. Performance isn't really an issue for either approach, and notifications can work, but with the complications identified in previous work here and other places, I think it makes sense to just use polling.